core/invoker/src/main/scala/whisk/core/containerpool/kubernetes/KubernetesClient.scala

```diff
@@ -0,0 +1,222 @@
+import scala.concurrent.duration.Duration
+import scala.concurrent.ExecutionContext
+import scala.concurrent.Future
+import scala.concurrent.blocking
+import scala.concurrent.duration._
+import scala.util.Failure
+import scala.util.Success
+import scala.util.Try
+
+import io.fabric8.kubernetes.client.ConfigBuilder
+import spray.json._
+import spray.json.DefaultJsonProtocol._
+import io.fabric8.kubernetes.client.DefaultKubernetesClient
+import io.fabric8.kubernetes.client.utils.URLUtils
+import okhttp3.Request
+
+/**
+ * Configuration for kubernetes client command timeouts.
+ */
+case class KubernetesClientTimeoutConfig(run: Duration, rm: Duration, inspect: Duration, logs: Duration)
+
+/**
+ * Serves as interface to the kubectl CLI tool.
+ *
+ * Be cautious with the ExecutionContext passed to this, as the
+ * calls to the CLI are blocking.
+ *
+ * You only need one instance (and you shouldn't get more).
+ */
+class KubernetesClient(
+  timeouts: KubernetesClientTimeoutConfig = loadConfigOrThrow[KubernetesClientTimeoutConfig](
+    "whisk.kubernetes.timeouts"))(executionContext: ExecutionContext)(implicit log: Logging, as: ActorSystem)
+    extends KubernetesApi
+    with ProcessRunner {
+  implicit private val ec = executionContext
+  implicit private val kubeRestClient = new DefaultKubernetesClient(
+    new ConfigBuilder()
+      .withConnectionTimeout(timeouts.logs.toMillis.toInt)
+      .withRequestTimeout(timeouts.logs.toMillis.toInt)
+      .build())
+
+  // Determines how to run kubectl. Failure to find a kubectl binary implies
+  // a failure to initialize this instance of KubernetesClient.
+  protected val kubectlCmd: Seq[String] = {
+    val alternatives = List("/usr/bin/kubectl", "/usr/local/bin/kubectl")
+
+    val kubectlBin = Try {
+      alternatives.find(a => Files.isExecutable(Paths.get(a))).get
+    } getOrElse {
+      throw new FileNotFoundException(s"Couldn't locate kubectl binary (tried: ${alternatives.mkString(", ")}).")
+    }
+
+    Seq(kubectlBin)
+  }
+
+  def run(image: String, name: String, environment: Map[String, String] = Map(), labels: Map[String, String] = Map())(
+    implicit transid: TransactionId): Future[ContainerId] = {
+    val environmentArgs = environment.flatMap {
+      case (key, value) => Seq("--env", s"$key=$value")
+    }.toSeq
+
+    val labelArgs = labels.map {
+      case (key, value) => s"$key=$value"
+    } match {
+      case Seq() => Seq()
+      case pairs => Seq("-l") ++ pairs
+    }
+
+    val runArgs = Seq(
+      "run",
+      name,
+      "--image",
+      image,
+      "--generator",
+      "run-pod/v1",
+      "--restart",
+      "Always",
+      "--limits",
+      "memory=256Mi") ++ environmentArgs ++ labelArgs
+
+    runCmd(runArgs, timeouts.run)
+      .map { _ =>
+        name
+      }
+      .map(ContainerId.apply)
+  }
+
+  def inspectIPAddress(id: ContainerId)(implicit transid: TransactionId): Future[ContainerAddress] = {
+    Future {
+      blocking {
```

Comment:
Is there a non-blocking client prodiving the features we need? (https://github.com/doriordan/skuber for example?)

Reply:
All the clients are going to bottom out with making http put/get against the same k8s REST API.  So, I'm not sure if it really matters in the end if we have the blocking explicit in our code or hide it in the bowels of 3rd party client instead.  

I'd bias towards picking a client based on expected long-term maintenance/support, expectation of it tracking k8s API changes, and license compatibility.

Reply:
Sure, that's why I added "providing the features we need". I'm all in for not just picking it for the sake of it. Still a non-blocking client is preferable over a blocking one (if it exists). This might matter if you want to scale the invoker to more than the 16 containers that we scale it today.

Reply:
We're planning to scale each Invoker to way more than 16 containers, but we haven't done large-scale load testing of this yet. I'd say we're targeting at least 100 or more containers per Invoker.

Reply:
Circling back on the client question.  The fabric8io client appears to lack support for specifying pod affinity.  Skimming through the source code, I see no support for affinity (and several open issues on GitHub asking for affinity support). The skuber client looks like it does have affinity support.  I think affinity is a required feature (lets us tell kube to restrict creation of user action containers to designated invoker nodes).

Reply:
With regards to the non-blocking, I've solved this down stream and am working on finishing testing. The fabric8 uses okhttp, which has an async mode. I've migrated to a custom `Source`, using the async callbacks, to eliminate most of the cruft and any blocking code. (I'm not sure about the pod affinity however)

Reply:
cool!  Maybe we can help out by adding the pod affinity support downstream as well.  I don't think it should be too hard.

Reply:
After reviewing, with regards to this section of code, I can't see a way with the fabric8 java client to *not* block here. Even if you avoid the utility method for waiting that we use in this code, every layer all the way down the client blocks at some point for it. and since this isn't able to just be a simple REST call like the log fetch, I can't refactor it to be async. The blocking is wrapped and marked inside the future, however, and will be handled safely.

Reply:
Makes sense to me.  Fundamentally we need to wait somehow until the pod is ready before we can inspect its Status object to get the PodIP and similar info.